app/controllers/lti_assignment_users_controller.rb

```diff
@@ -0,0 +1,221 @@
+  def pre_process_tenant
+    error_code, message = nil
+    oauth_params = OAuth::OAuthProxy::OAuthRequest.parse_authorization_header request.authorization
+    oauth_consumer_key = oauth_params[:oauth_consumer_key] || params['oauth_consumer_key']
+
+    # OAuth check here
+    tool_provider_registry = Rails.application.config.tool_provider_registry
+
+    key = oauth_consumer_key
+    unless key
+      error_code = "LTI_INVALID_REQUEST"
+      message = "Improper LTI context: LTI Consumer key is missing or not valid!"
+      return [error_code, message]
+    end
+    @tenant = Lti_Tenant.where(:tenant_key => key).first
+    @registration = Lti2Tp::Registration.where(:tenant_id => @tenant.id).first
+    unless @registration
+      error_code = "LTI_INVALID_CONSUMER"
+      message = "Tool registration not found. Please register Expertiza in your LMS before invoking the request."
+      return [error_code, message]
+    end
+
+    request.parameters['_tenant_id'] = @tenant.id
+
+    tool_proxy = JsonWrapper.new(@registration.tool_proxy_json)
+    secret = @tenant.secret
+
+    unless tool_provider_registry.relaxed_oauth_check == 'true'
+      request_wrapper = OAuthRequest.create_from_rack_request request
+      (is_success, signature_base_string) = request_wrapper.verify_signature? secret, Rails.application.config.nonce_cache
+      unless is_success
+        puts "TP Secret: #{secret}"
+        puts "TP Signed Request: #{request_wrapper.signature_base_string}"
+        (redirect_to redirect_url("Invalid signature")) and return
+      end
+      request[:lti_context] = request.parameters
+    end
+
+    # LTI conformance
+    # Perform extra LTI type checks only for launch
+    if params['lti_message_type']  == 'basic-lti-launch-request'
+      if params.has_key?('lti_message_type')
+        lti_message_type = params['lti_message_type']
+        unless ['basic-lti-launch-request', 'ToolProxyRegistrationRequest', 'ToolProxyReregistrationRequest'].include?(lti_message_type)
+          (redirect_to redirect_url("Invalid lti_message_type: #{lti_message_type}")) and return
+        end
+      else
+        (redirect_to redirect_url("Missing lti_message_type")) and return
+      end
+
+      if params['lti_message_type'] == 'basic-lti-launch-request'
+        request.request_parameters['normalized_role'] = normalize_role(params['roles'])
+        unless params.has_key?('resource_link_id')
+          (redirect_to redirect_url("Missing resource link id")) and return
+        end
+
+        if params.has_key?('lti_version')
+          lti_version = params['lti_version']
+          unless ['LTI-1p0', 'LTI-2p0'].include?(lti_version)
+            (redirect_to redirect_url("Invalid lti_version: #{lti_version}")) and return
+          end
+        else
+          (redirect_to redirect_url("Missing lti_version")) and return
+        end
+      end
+    end
+
+    return [200, nil, nil]
```

Comment:
A constant like "200" needs to be defined as a constant and explained in a comment.